besskge/scoring.py

```diff
@@ -763,20 +769,22 @@ def boxe_score(
         width_ht = width_ht / torch.clamp(
             torch.exp(
                 torch.mean(
-                    torch.log(torch.clamp(width_ht, min=self.soft)),
+                    torch.log(torch.clamp(width_ht, min=self.eps)),
                     dim=-1,
                     keepdim=True,
                 )
             ),
-            min=self.soft,
+            min=self.eps,
         )
         # Rescale by box_size
         width_ht = width_ht * (
             torch.tensor(1.0, dtype=torch.float32, device=width_ht.device)
             + torch.nn.functional.elu(box_size.unsqueeze(-1))
         )
         if self.apply_tanh:
-            width_ht = torch.tanh(width_ht)
+            width_ht = torch.tanh(width_ht) * torch.tensor(
```

Comment:
I'm probably missing something, but is this not different from PyKEEN where tanh is applied to the box edges?

Reply:
Correct, mathematically it's not exactly the same implementation. I would like to avoid having to materialize the box edges if possible: when not using `tanh` it's simple to do (while pykeen has wasted computations there), here it's a bit of a workaround. In the tests I ran it still seems to perform similarly to pykeen, although if we want to have the exact same results we would need to do it differently.

Reply:
Would it be closer to PyKeen if `in_or_out` would always be computed pre-tanh? Due to monotonicity of tanh that should be equivalent to the case where tanh is applied to the box edges, whereas applying tanh to the box width can change which points fall into the box

Reply:
> So the factor 2 comes in with the approximation `tanh(x)~x` for small x?

The factor of 2 was just so the final range of `widht_ht` was [0,2], i.e. the same of what we would have if we took the difference of the tanh of the upper and lower edge.

> Would it be closer to PyKeen if `in_or_out` would always be computed pre-tanh?  Due to monotonicity of tanh that should be equivalent to the case where tanh is applied to the box edges, whereas applying tanh to the box width can change which points fall into the box

That is an excellent idea and I gave it a try with high expectations, but turns out to be actually slightly detrimental on final MRR (not sure why).

In the end I resorted to compute the upper and lower edges and do tanh like pykeen: impact on throughput is negligible and it does perform a bit better on MRR. I also moved some computations inside the `boxe_score` auxiliary function to minimize similar code across `score_triple, score_head, score_tail`. Hopefully this is the last iteration, but do tell if you see anything that could be still done better!